lib/router/src/cache.rs

```diff
@@ -122,14 +113,20 @@ where
     ///
     /// Polls the underlying `DelayQueue`. When elements are returned from the
     /// queue, remove the associated key from `values`.
-    fn poll_purge(&mut self) -> Poll<(), Error> {
-        while let Some(expired) = try_ready!(self.expirations.poll()) {
-            debug!("expiring an item from the cache");
-            self.values.remove(expired.get_ref());
+    fn poll_purge(&mut self) -> Poll<(), ()> {
+        loop {
+            match try_ready!(self
+                .expirations
+                .poll()
+                .map_err(|e| unreachable!("expiration must not fail: {}", e)))
+            {
+                None => return Ok(Async::NotReady),
```

Comment:
If `poll_purge` returns `NotReady` here, will the lock we acquired in `poll` [here](https://github.com/linkerd/linkerd2-proxy/pull/286/files#diff-230615b3316b2a089b0d6563c3f74944R144) get released? Or, will we continue to loop within this `poll_purge` and always hold on to the lock that we acquired?

Reply:
Looking at this, I'm not immediately sure if this returning `NotReady` would pop up to the `poll` where the lock was acquired. And if that is not happening, then that would be an issue.

Reply:
I'm not sure I understand the question... Given the future impl:

```rust
impl<K, V> Future for PurgeCache<K, V>
where
    K: Clone + Eq + Hash,
    V: Clone,
{
    type Item = ();
    type Error = ();

    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        let mut cache = try_ready!(Ok(self.0.poll_lock()));
        cache.poll_purge()
    }
}
```

When poll_purge returns not ready, the lock is released (as the guard is dropped at the end of the scope).

Reply:
Ok cool thanks for talking this through I’m following what’s going on and it makes sense.

Reply:
However, as it is, the purge future may never complete when the cache is dropped. I have a small modification to this branch that will fix this.

Reply:
Actually, since this doesn't pose a current problem for the proxy, let's defer this (it's a little more complicated than i initially thought).